toolchains/prebuilt_toolchains.py

```diff
@@ -0,0 +1,370 @@
+    toolchain = ":ninja_tool",
+    toolchain_type = "@rules_foreign_cc//tools/build_defs:ninja_toolchain",
+)
+
+\"\"\"
+
+# buildifier: disable=unnamed-macro
+def prebuilt_toolchains(cmake_version, ninja_version):
+    \"\"\"Register toolchains for pre-built cmake and ninja binaries
+
+    Args:
+        cmake_version (string): The target cmake version
+        ninja_version (string): The target ninja-build version
+    \"\"\"
+    _cmake_toolchains(cmake_version)
+    _ninja_toolchains(ninja_version)
+    _make_toolchains()
+
+def _cmake_toolchains(version):
+{cmake_definitions}
+
+def _ninja_toolchains(version):
+{ninja_definitions}
+
+def _make_toolchains():
+{make_definitions}
+"""
+
+
+def get_cmake_definitions() -> str:
+    """Define a set of repositories and calls for registering `cmake` toolchains
+
+    Returns:
+        str: The Implementation of `_cmake_toolchains`
+    """
+
+    archives = []
+
+    for version in CMAKE_VERSIONS:
+        major, minor, _patch = version.split(".")
+
+        version_archives = []
+        version_toolchains = []
+
+        minor_version = "{}.{}".format(major, minor)
+        for line in urllib.request.urlopen(CMAKE_SHA256_URL_TEMPLATE.format(minor=minor_version, full=version)).readlines():
+            line = line.decode("utf-8").strip("\n ")
+
+            # Only take tar and zip files. The rest can't be easily decompressed.
+            if not line.endswith(".tar.gz") and not line.endswith(".zip"):
+                continue
+
+            # Only include the targets we care about.
+            plat_target = None
+            for target in CMAKE_TARGETS.keys():
+                if target in line:
+                    plat_target = target
+                    break
+
+            if not plat_target:
+                continue
+
+            sha256, file = line.split()
+            name = file.replace(".tar.gz", "").replace(".zip", "")
+            bin = "cmake.exe" if "win" in file.lower() else "cmake"
+
+            if "Darwin" in file or "macos" in file:
+                prefix = name + "/CMake.app/Contents"
+            else:
+                prefix = name
+
+            version_archives.append(
+                REPO_DEFINITION.format(
+                    name=name,
+                    sha256=sha256,
+                    prefix=prefix,
+                    url=CMAKE_URL_TEMPLATE.format(
+                        full=version,
+                        file=file
+                    ),
+                    build="cmake",
+                    template="_CMAKE_BUILD_FILE",
+                    bin=bin,
+                    exec_compatible_with=CMAKE_TARGETS[plat_target],
+                )
+            )
+            version_toolchains.append("@{}//:cmake_toolchain".format(name))
+
+        archives.append("\n".join(
+            [
+                "    if \"{}\" == version:".format(version),
+            ] + [indent(archive, " " * 8) for archive in version_archives])
```

Comment:
The version_archives expansion doesn't need to be protected by the if. There's no harm in defining all the repositories unconditionally and this allows an end user to refer to these to setup their own toolchain if necessary using these repositories e.g. they may want to setup a toolchain with an additional exec constraint so that they can select an older version of a tool for a particular build if the newest version doesn't work for them.

Reply:
Would you be ok with leaving this as a later improvement?

Reply:
moving conversation from https://github.com/bazelbuild/rules_foreign_cc/pull/508#discussion_r576669270 here